dashboard/app/models/pd/application/teacher1819_application.rb

```diff
@@ -426,6 +427,36 @@ def check_idempotency
       Pd::Application::Teacher1819Application.find_by(user: user)
     end
 
+    def workshop
+      return unless regional_partner
+
+      # If this application is associated with a G3 partner who in turn is
+      # associated with a specific teachercon, return the workshop for that
+      # teachercon
+      if regional_partner.group == 3
+        teachercon = get_matching_teachercon(regional_partner)
+        if teachercon
+          return find_teachercon_workshop(course, teachercon, 2018)
+        end
+      else
+
+      # Default to just assigning whichever of the partner's workshops is
+      # scheduled to start first. We expect to hit this case for G1 and G2
+      # partners, and for any G3 partners without an associated teachercon
+      regional_partner.workshops.order_by_scheduled_start.first
+    end
+
+    def assign_workshop!
+      return unless workshop
+
+      ::Pd::Enrollment.new(
+        workshop: workshop,
+        user: user,
+        full_name: user.name,
+        email: user.email,
+      )
```

Comment:
Do we want this "assignment" to only be tracked by the external enrollment? As it stands this is not editable / retrievable here. We can call `workshop` again, but if the enrollment changes, it won't reflect the actual assigned workshop. 

I'd like to see this be idempotent and retrievable. I suggest changing `workshop` (maybe rename to `assigned_workshop`?) to inspect enrollments and return the one that is enrolled if any, rather than first, with first as a default when none are enrolled. Then, with `assign_workshop!` (maybe renamed `enroll_in_assigned_workshop`?), we should at least use `find_or_create_by!` rather than new. 

We could make it even more flexible for reassignment by allowing an optional argument for the specific workshop to be assigned and first deleting any other relevant enrollments (same course and subject).


What do you think?

Reply:
I agree that this should be explicitly set rather than inferred.

I think the answer to how best to build this is actually going to kind of depend on the answer to whether we need to filter by course and/or subject. Is it the case that all the workshops we will want to be assignable to a teacher application are going to share some data, or do we want to allow partners to assign whatever workshop they want? I think it's gonna be the latter, since some of the G3-associated teachers are going to end up going to Local Summer Workshops and we'll want to handle those differently.

If we want that flexibility, it seems to me that we'll want an actual direct relation from application to workshop or enrollment, and `assign_workshop!` is refactored to something like `self.enrollment = Pd::Enrollment.new(...) unless self.enrollment`

Reply:
The teachercon vs. local summer thing does muddy this up a bit, but the course should always be the applied-for course (CSD vs. CSP), exposed as `course` in the application, though it will need to be translated to be used in a workshop query (see https://github.com/code-dot-org/code-dot-org/blob/7b2913fad4ef41e6c607ceb456fb1e06ebfc880d/dashboard/app/models/pd/teacher_application.rb#L25).

I suppose subject could be the set of
```
[
  Pd::Workshop::SUBJECT_TEACHER_CON, 
  Pd::Workshop::SUBJECT_FIT
]
```